runtime/master/src/main/java/org/apache/nemo/runtime/master/scheduler/SimulationScheduler.java

```diff
@@ -0,0 +1,579 @@
+import org.apache.nemo.runtime.common.message.MessageUtils;
+import org.apache.nemo.runtime.common.metric.JobMetric;
+import org.apache.nemo.runtime.common.metric.Metric;
+import org.apache.nemo.runtime.common.plan.PhysicalPlan;
+import org.apache.nemo.runtime.common.plan.PlanRewriter;
+import org.apache.nemo.runtime.common.plan.Stage;
+import org.apache.nemo.runtime.common.plan.Task;
+import org.apache.nemo.runtime.common.state.StageState;
+import org.apache.nemo.runtime.common.state.TaskState;
+import org.apache.nemo.runtime.master.BlockManagerMaster;
+import org.apache.nemo.runtime.master.PlanAppender;
+import org.apache.nemo.runtime.master.PlanStateManager;
+import org.apache.nemo.runtime.master.metric.MetricStore;
+import org.apache.nemo.runtime.master.resource.DefaultExecutorRepresenter;
+import org.apache.nemo.runtime.master.resource.ExecutorRepresenter;
+import org.apache.reef.annotations.audience.DriverSide;
+import org.apache.reef.driver.context.ActiveContext;
+import org.apache.reef.driver.evaluator.EvaluatorDescriptor;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.inject.Inject;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * Scheduler for simulating an execution, not controlled by the runtime master.
+ */
+@DriverSide
+@NotThreadSafe
+public final class SimulationScheduler implements Scheduler {
+  private static final Logger LOG = LoggerFactory.getLogger(SimulationScheduler.class.getName());
+
+  /**
+   * Run-time optimizations.
+   */
+  private final PlanRewriter planRewriter;
+
+  /**
+   * Components related to scheduling the given plan.
+   */
+  private TaskDispatcher taskDispatcher;
+  private final PendingTaskCollectionPointer pendingTaskCollectionPointer;
+  private ExecutorRegistry executorRegistry;
+  private PlanStateManager planStateManager;
+  private final ExecutorService serializationExecutorService; // Executor service for scheduling message serialization.
+  private final BlockManagerMaster blockManagerMaster;
+  private final MetricStore actualMetricStore;
+  private MetricStore metricStore;
+  private CountDownLatch metricCountDownLatch;
+
+  private final SchedulingConstraintRegistry schedulingConstraintRegistry;
+  private final SchedulingPolicy schedulingPolicy;
+  private final String resourceSpecificationString;
+  private final String dagDirectory;
+
+  private final Map<String, SimulatedTaskExecutor> simulatedTaskExecutorMap;
+
+  /**
+   * The below variables depend on the submitted plan to execute.
+   */
+  private List<List<Stage>> sortedScheduleGroups;
+
+  @Inject
+  private SimulationScheduler(final PlanRewriter planRewriter,
```

Comment:
The purpose of this class is to somehow 'mirror' the behaviors of batch scheduler with similar code. Does this mean this class should be modified when the batch scheduler is modified? Can you leave a comment about this so we don't forget later when we do edit the batch scheduler?

Reply:
I'll leave comments accordingly.